### ISSUE TITLE
Forge cheatcode for recording keccak preimages

### ISSUE THREAD

**Issue by kopy-kat** [foundry-rs/foundry#6837]
### Component

Forge

### Describe the feature you would like

A set of cheatcodes, similar to `record/accesses` and `recordLogs/getRecordedLogs` but for logging and returning the preimages used by the `keccak256` opcode.

Use case: related to https://github.com/foundry-rs/foundry/issues/6704, this is required to perform the spec validations for ERC-4337, which only allows reading/writing slots that are associated with a specific address in some circumstances. `record/access` is sufficient in cases in which this address is used directly as a slot value, but not in cases in which it is hashed together with another value, such as when using a mapping. It is also not sufficient to try to brute force mappings with low indexes, because this will not work for nested mappings.

### Suggestion
- `function recordPreImages() external;` - which starts recording all keccak preimages sliced out from memory
- `function getRecordedPreImages() external returns (bytes[] memory);` - which returns all recorded preimages

**Comment by kopy-kat** [comment]
I'm also happy to take out a pr for this if it looks good

**Comment by mds1** [comment]
Does https://github.com/foundry-rs/foundry/pull/5123 (https://github.com/foundry-rs/foundry/pull/4710) handle your use case?

**Comment by kopy-kat** [comment]
As far as I can tell no, because it assumes that I know the root slot of the mapping which is not the case (correct me if I'm wrong). 

I could loop over small indexes and try every slot but this would be quite hacky. However, one way this could work for my use case relatively easily is to change `stopMappingRecording` or add a new function that returns an array of all recorded mappings (similar to `stopAndReturnStateDiff`). Wdyt?

**Comment by kopy-kat** [comment]
Also fyi I don't think the changes from https://github.com/foundry-rs/foundry/pull/5123 are in the foundry book yet

**Comment by mds1** [comment]
@kopy-kat I see, thanks. Can you expand on the 4337 use case? I'm not too familiar with the validation details but it sounds like you still only need preimages for keccaks which precede an SSTORE (mapping write, solidity generated hash) as opposed to for arbitrary `keccak256(...)` invocations in solidity? Your suggestion of a new method to return the required data, instead of a new set of cheats, feels like the right direction, but would love help fleshing out the UX

@clouds56 any chance you could open a PR to https://github.com/foundry-rs/book to add docs for those cheats?

**Comment by drortirosh** [comment]
Two small suggestions:
1. Since it records the preimage of keccak explicitly, it is better to name it `recordKeccakPreImages()`
2. A small performance optimization is to provide "max preimageSize" parameter, to avoid recording (and returning) large preimages, of entire contracts (technically, with erc-4337 we only need to find mapping indices, which are keccak over blocks of 64 bytes)

**Comment by kopy-kat** [comment]
@mds1 after some prototyping, I've managed to get this to work for the ERC-4337 usecase with the `getMappingKeyAndParentOf` cheatcode that you mention above, so closing this issue - thanks for the help!

**Comment by drortirosh** [comment]
actually, while this cheatcode is great, we do need for erc4337 the revese cheatcode: return the raw binary data that was keccak'ed
eg.
```
getArrayOfMappingKeyAndParent(address) returns (bytes[] memory)
```
The existing getMappingKeyAndParentOf allows us to tell for a given slot access (retrieved from stopAndReturnStateDiff) , how it is used in a mapping such as `mapping (address=>uint)`

However, if the slot reference is into a struct member (`mapping(address=>Struct)`) that is not the first one, then we have to do an "exhaustive search", trying different offsets.

The above suggested API (which merely exposes already collected data from `seen_sha3.get(address).values()`) Would simplify such query

**Comment by kopy-kat** [comment]
@mds1 happy to do this but to complete 4337 verification, https://github.com/foundry-rs/foundry/issues/6704 (or something similar) is also required (to check banned opcodes)

**Comment by kopy-kat** [comment]
@drortirosh the upside of the proposed cheatcode would just be that you can iterate over the bytes in the contract and check whether `slot - n` is in the array for every n, right? I'm not sure if that is a huge upside as oppposed to just doing a call to the cheatcode address for every n

**Comment by drortirosh** [comment]
tl;dr - I agree the upside is not huge, as this is a test framework (and not a production server) the actual performance is not that critical.

currently, to support structures of sizes up to (say) 128 words, we need to iterate for each storage access up to 128 times.
If we had access to the raw "keccak'ed" data, we'd need to iterate over this array times the number of possible addresses, which is up to 4. Also, the total size of this array is bounded by the total number of slots, but in practice it can be much smaller.

**Comment by mds1** [comment]
What's the consensus here, should we reopen this issue? It sounds like for full 4337 testing:
- This issue is a nice to have
- https://github.com/foundry-rs/foundry/issues/6704 is required

Is that correct? If so, for simplicity perhaps we keep this closed and reopen if performance with the existing method becomes an issue?

As a side note, it feels like there's a great opportunity for an `erc-4337-validation` Solidity testing library that abstracts these checks for users

**Comment by kopy-kat** [comment]
Yes, that is correct. I am actually working on a lib right now: https://github.com/rhinestonewtf/erc4337-validation :)
